hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/ClientGSIContext.java

```diff
@@ -77,11 +83,45 @@ public void updateResponseState(RpcResponseHeaderProto.Builder header) {
   @Override
   public synchronized void receiveResponseState(RpcResponseHeaderProto header) {
     if (header.hasRouterFederatedState()) {
-      routerFederatedState = header.getRouterFederatedState();
+      routerFederatedState = mergeRouterFederatedState(header.getRouterFederatedState());
     } else {
       lastSeenStateId.accumulate(header.getStateId());
     }
   }
+  /**
+   * Utility function to parse routerFederatedState field in RPC headers.
+   */
+  public static Map<String, Long> getRouterFederatedStateMap(ByteString byteString) {
+    if (byteString == null) {
+      return Collections.emptyMap();
+    }
+
+    RouterFederatedStateProto federatedState;
+    try {
+      federatedState = RouterFederatedStateProto.parseFrom(byteString);
+    } catch (InvalidProtocolBufferException e) {
+      throw new RuntimeException(e);
+    }
+    return federatedState.getNamespaceStateIdsMap();
```

Comment:
Might be cleaner to do this inside of the try and move the declaration of the var too.
Then you can return Collections.emptyMap(); at the end